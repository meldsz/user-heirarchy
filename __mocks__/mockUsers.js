export const mockUsers = {
  secrets: {
    ABCD: 123,
    GHYI: 567,
  },
  users: [
    {
      email: "john.smith@xtreet.tvl",
      firstName: "John",
      id: 123,
      lastName: "Smith",
      password: "asak",
    },
    {
      email: "matthew.bartolacci@btthew.net",
      firstName: "Matthew",
      id: 567,
      lastName: "Bartolacci",
      password: "x6Ldx7Ye",
      photo: "https://test.jpg",
    },
    {
      email: "jane.smith@xtreet.tvl",
      firstName: "Jane",
      id: 777,
      managerId: 567,
      lastName: "Smith",
      password: "uiu",
    },
    {
      email: "jane.doe@xtreet.tvl",
      id: 444,
      password: "iur",
    },
  ],
};
