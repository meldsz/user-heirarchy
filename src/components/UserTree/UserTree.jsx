import { useEffect, useState } from "react";
import { fetchUsers } from "../../common/api";
import { UserNode } from "../UserNode/UserNode";

export const UserTree = () => {
  const [tree, setTree] = useState([]);

  useEffect(() => {
    const getAllUsers = async () => {
      const allUsers = await fetchUsers();
      const userTree = buildUserTree(allUsers?.users);
      setTree(userTree);
    };
    getAllUsers();
  }, []);

  const buildUserTree = (allUsers) => {
    const tree = {};
    const root = [];
    allUsers.forEach((user) => (tree[user.id] = { ...user, children: [] }));
    allUsers.forEach((user) => {
      if (user?.managerId) {
        tree[user?.managerId].children.push(tree[user?.id]);
      } else {
        root.push(tree[user?.id]);
      }
    });

    return root;
  };

  return (
    <div className="tree-container">
      {tree && tree.map((user) => <UserNode key={user.id} user={user} />)}
    </div>
  );
};
