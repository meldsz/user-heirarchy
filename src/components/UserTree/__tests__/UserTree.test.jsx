import { mockUsers } from "../../../../__mocks__/mockUsers";
import { UserTree } from "../UserTree";
import { act, fireEvent } from "@testing-library/react";
import { testRender } from "../../../common/tests/helper";

jest.mock("../../../common/api", () => ({
  fetchUsers: jest.fn().mockResolvedValue(mockUsers),
}));

describe("<UserTree/>", () => {
  it("should display user tree correctly", async () => {
    let wrapper;
    await act(async () => {
      wrapper = testRender(<UserTree />);
    });
    const { queryAllByTestId, asFragment } = wrapper;
    fireEvent.click(queryAllByTestId("toggle")[1]);
    expect(asFragment()).toMatchSnapshot();
  });
});
