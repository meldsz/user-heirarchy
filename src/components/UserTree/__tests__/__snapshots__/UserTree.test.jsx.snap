// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<UserTree/> should display user tree correctly 1`] = `
<DocumentFragment>
  <div>
    <div
      data-testid="user-node-container"
    >
      <div
        class="user-container"
      >
        <button
          aria-label="Expand"
          class="toggle"
          data-testid="toggle"
        >
          <strong>
            -
          </strong>
        </button>
        <div
          class="user-badge"
        >
          JS
        </div>
        <div
          class="user-details"
        >
          <div>
            John Smith
          </div>
          <div>
             john.smith@xtreet.tvl
          </div>
        </div>
      </div>
    </div>
    <div
      data-testid="user-node-container"
    >
      <div
        class="user-container"
      >
        <button
          aria-label="Collapse"
          class="toggle"
          data-testid="toggle"
        >
          +
        </button>
        <img
          class="user-badge"
          data-testid="user-badge"
          src="https://test.jpg"
        />
        <div
          class="user-details"
        >
          <div>
            Matthew Bartolacci
          </div>
          <div>
             matthew.bartolacci@btthew.net
          </div>
        </div>
      </div>
      <div
        class="user-children"
      >
        <div
          data-testid="user-node-container"
        >
          <div
            class="user-container"
          >
            <button
              aria-label="Expand"
              class="toggle"
              data-testid="toggle"
            >
              <strong>
                -
              </strong>
            </button>
            <div
              class="user-badge"
            >
              JS
            </div>
            <div
              class="user-details"
            >
              <div>
                Jane Smith
              </div>
              <div>
                 jane.smith@xtreet.tvl
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      data-testid="user-node-container"
    >
      <div
        class="user-container"
      >
        <button
          aria-label="Expand"
          class="toggle"
          data-testid="toggle"
        >
          <strong>
            -
          </strong>
        </button>
        <div
          class="user-badge"
        >
          N/A
        </div>
        <div
          class="user-details"
        >
          <div>
             
          </div>
          <div>
             jane.doe@xtreet.tvl
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
